File: LeetCode/0167. Two Sum II - Input array is sorted/Solution.java
package io.innofang;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by Inno Fang on 2018/2/6.
 */
class Solution {

    /**
     * 16 / 16 test cases passed.
     * Status: Accepted
     * Runtime: 5 ms
     *
     * @param numbers
     * @param target
     * @return
     */
    public int[] twoSum(int[] numbers, int target) {
        Map<Integer, Integer> store = new HashMap<>();
        for (int i = 0; i < numbers.length; i++) {
            int num = numbers[i];
            if (store.containsKey(target - num)) {
                return new int[]{store.get(target - num) + 1, i + 1};
            }
            store.put(num, i);
        }
        return new int[]{};
    }
}

File: LeetCode/0167. Two Sum II - Input array is sorted/solution.cpp
/**
 * 21 / 21 test cases passed.
 * Runtime: 24 ms
 * Memory Usage: 15.3 MB 
 */
class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        int l = 0, r = numbers.size() - 1;
        while (l <= r) {
            int sum = numbers[l] + numbers[r];
            if (sum == target) return {l + 1, r +1};
            else if (sum > target) -- r;
            else ++ l;
        }
        return {-1, -1};
    }
};
